source/s3_auto_ranged_put.c

```diff
@@ -1231,10 +1237,19 @@ static void s_s3_prepare_upload_part_finish(struct aws_s3_prepare_upload_part_jo
             aws_string_c_str(auto_ranged_put->upload_id));
 
     } else {
+        struct aws_byte_buf **checksum_buf_pointer = NULL;
         aws_array_list_get_at_ptr(
-            &auto_ranged_put->synced_data.encoded_checksum_list, (void **)&checksum_buf, request->part_number - 1);
+            &auto_ranged_put->synced_data.encoded_checksum_list,
+            (void **)&checksum_buf_pointer,
+            request->part_number - 1);
         /* Clean up the buffer in case of it's initialized before and retry happens. */
-        aws_byte_buf_clean_up(checksum_buf);
+        checksum_buf = *checksum_buf_pointer;
+        if (checksum_buf) {
+            aws_byte_buf_clean_up(checksum_buf);
+        } else {
+            checksum_buf = aws_mem_calloc(meta_request->allocator, 1, sizeof(struct aws_byte_buf));
+            *checksum_buf_pointer = checksum_buf;
+        }
```

Comment:
this code is accessing `synced_data.encoded_checksum_list` without holding the lock

also, it's so damned confusing, it took me 20+ minutes to wrap my head around why we need to sometimes clean_up(), and other times allocate (but not aws_byte_buf_init())

I'm torn between suggesting comments that explaining what's going on here

or suggesting ways to make this code less confusing (stop allocating the byte-buf that the weird stream will initialize later)

maybe we can move this 1 block of code around to make more sense:
1) delete everything here.  `s_s3_prepare_upload_part_finish()` should not touch `synced_data.encoded_checksum_list`
2) move the `aws_mem_calloc(..., sizeof(struct aws_byte_buf))` up into the function above this, where we're already holding a lock to resize `synced_data.encoded_checksum_list`. But instead of setting NULL, set the allocated byte-buf
3) move the `aws_byte_buf_clean_up(checksum_buf);` up into `s_s3_prepare_upload_part()` in the same else block with the comment `/* Not the first time preparing request (e.g. retry).`

Reply:
+1, but i think we should address it in 2 steps. let fix the issue first with the way it currently works and lets discuss how to refactor it better and we can do that in follow up pr

Reply:
yeah, me and Dengke just had a chat and arrived at the same conclusion.
I can revisit this when I let the user pass in checksum